### PR TITLE
add IA create/update

### PR DESCRIPTION
**Pull request by SergiyButenko8**

### CODE REVIEW

File: app/models/category.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class Category < ApplicationRecord
+  acts_as_list add_new_at:  :top
```
Comment (OrestOstashuk) [inline app/models/category.rb:2]: run rubocop
two spaces

File: app/models/sub_category.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class SubCategory < ApplicationRecord
+  acts_as_list scope: :category, add_new_at:  :top
```
Comment (OrestOstashuk) [inline app/models/sub_category.rb:2]: same as above ^

File: app/controllers/account/admin/categories_controller.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+module Account::Admin
+  class CategoriesController < AdminBaseController
+    before_action :set_category, only: %i[edit update destroy change_cat_visibility move_position]
+    before_action :set_page_title
+    before_action :authorize_admin_user
+    def index
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/categories_controller.rb:8]: new line pls between def and before_...

File: app/controllers/account/admin/categories_controller.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        end
+      else
+        flash[:alert] = @category.errors.full_messages.to_sentence
+        redirect_to account_admin_categories_path
+      end
+    end
+
+    def change_cat_visibility
+      @categories = Category.all.order(id: :desc)
+      if (@category.hidden? ? "true" : "false") != (category_params[:hidden]) && @category.update(category_params)
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/categories_controller.rb:45]: `@category.hidden?.to_s` can be so

File: app/controllers/account/admin/categories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        respond_to do |format|
+          format.js do
+            flash.now[:notice] = @category.hidden? ? "This category is hidden now" : "This category is visible now"
+          end
+        end
+      else
+        flash[:alert] = "Something wend wrong"
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/categories_controller.rb:52]: went*

File: app/controllers/account/admin/categories_controller.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      head :ok
+    end
+
+    def destroy
+      if @category.destroy
+        respond_to do |format|
+          format.js { flash.now[:notice] = "Category deleted successfully" }
+        end
+      else
+        flash[:alert] = "Something wend wrong"
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/categories_controller.rb:68]: went*

File: app/controllers/account/admin/categories_controller.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      end
+    end
+
+    private
+
+    def category_params
+      params.require(:category).permit(:label, :hidden)
+    end
+
+    def set_category
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/categories_controller.rb:79]: You dont need this method at all
you can have just
```
def category
@category || = ...
```
and call it where you need
-1 before action.

File: app/controllers/account/admin/sub_categories_controller.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
+# frozen_string_literal: true
+
+module Account::Admin
+  class SubCategoriesController < AdminBaseController
+    before_action :set_sub_category, only: %i[edit update destroy change_sub_visibility move_to_category move_position]
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/sub_categories_controller.rb:5]: same as for categories

File: app/controllers/account/admin/sub_categories_controller.rb [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    def sub_category_params
+      params.require(:sub_category).permit(:label, :hidden)
+    end
+
+    def set_sub_category
+      @sub_category = category.sub_categories.find(params[:id])
+    end
+
+    def category
+      @category = Category.find(params[:category_id])
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/sub_categories_controller.rb:98]: so do you need @category ?
or memoize here ?
or remove `@category =` and leave just` Category.find(params[:category_id])`

File: app/controllers/account/admin/teams_controller.rb [hunk 10]
```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+module Account::Admin
+  class TeamsController < AdminBaseController
+    before_action :set_team, only: %i[edit update destroy change_team_visibility move_to_sub_category move_position]
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/teams_controller.rb:5]: same )

File: app/controllers/account/admin/teams_controller.rb [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    def team_params
+      params.require(:team).permit(:label, :hidden, :move_to_sub_cat_id)
+    end
+
+    def set_team
+      @team = sub_category.teams.find(params[:id])
+    end
+
+    def category
+      @category = Category.find(params[:category_id])
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/teams_controller.rb:93]: same as i asked above

File: app/controllers/account/admin/teams_controller.rb [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    end
+
+    def sub_category
+      @sub_category = category.sub_categories.find(params[:sub_category_id])
+    end
+
+    def categories
+      @categories = Category.includes(:sub_categories).all
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/teams_controller.rb:101]: same as i asked above

File: app/controllers/account/admin/users_controller.rb [hunk 13]
```diff
@@ -5,6 +5,8 @@ class UsersController < AdminBaseController
     before_action :set_user, except: %i[index]
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/users_controller.rb:5]: same
we can ommit this before action

File: app/handlers/team_move_to_handler.rb [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  private
+
+  def move_to_sub_category
+    @move_to_sub_category ||= SubCategory.find(@sub_category_id)
+  end
+
+  def current_team
+    @current_team = SubCategory.find(@team.sub_category_id)
```
Comment (OrestOstashuk) [inline app/handlers/team_move_to_handler.rb:25]: same
why do you need variable `@current_team` ?

File: app/handlers/team_move_to_handler.rb [hunk 15]
```diff
@@ -0,0 +1,27 @@
+class TeamMoveToHandler
+  def initialize(team, params)
+    @team = team
```
Comment (OrestOstashuk) [inline app/handlers/team_move_to_handler.rb:3]: you can create attr readers for params
private if needed

File: app/views/account/admin/categories/_show.html.slim [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+    - if category.hidden?
+      .badge.rounded-pill.hidden-badge hidden
+    = link_to "#{category.label}", account_admin_category_sub_categories_path(category), remote: true, class: ("cat-btn")
+  button type="button" class="btn dropdown-toggle dropdown-toggle-split" data-bs-toggle="dropdown" aria-expanded="false"
+    i.fas.fa-ellipsis-h style="color: #d4d4d4"
+  ul.dropdown-menu.py-0
+    li
+      = link_to "Edit", edit_account_admin_category_path(category), remote: true, class: "dropdown-item"
+      = link_to "Delete", account_admin_category_path(category), method: :delete, data: { confirm: "Are you sure you want to delete this subcategory?", remote: true }, class: "dropdown-item delete-cat"
+      = link_to "#{category.hidden? ? "Show" : "Hide"}", change_cat_visibility_account_admin_category_path(category, category: { hidden: !category.hidden }), remote: true, method: :put, class: "dropdown-item"
```
Comment (OrestOstashuk) [inline app/views/account/admin/categories/_show.html.slim:12]: is it working ?)

File: db/migrate/20210622072443_add_position_to_sub_categories.rb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+class AddPositionToSubCategories < ActiveRecord::Migration[6.1]
+  def change
+    add_column :sub_categories, :position, :integer, default: 1
```
Comment (OrestOstashuk) [inline db/migrate/20210622072443_add_position_to_sub_categories.rb:3]: why 3 separate migrations for sub categories and teams ?

File: app/controllers/account/admin/categories_controller.rb [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        flash[:alert] = @category.errors.full_messages.to_sentence
+        redirect_to account_admin_categories_path
+      end
+    end
+
+    def update
+      @categories = Category.all.order(position: :asc)
+      if category.update(category_params)
+        respond_to do |format|
+          format.js { flash.now[:notice] = "Category updated successfully" }
```
Comment (OrestOstashuk) [inline app/controllers/account/admin/categories_controller.rb:37]: all srtings can be moved to locales

Comment (chu4) [reply app/controllers/account/admin/categories_controller.rb:37]: we weren't doing I18n yet

File: app/controllers/account/admin/sub_categories_controller.rb [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+        end
+      else
+        flash[:alert] = sub_category.errors.full_messages.to_sentence
+        redirect_to account_admin_categories_path
+      end
+    end
+
+    def change_sub_visibility
+      @sub_categories = category.sub_categories.order(position: :asc)
+      if sub_category.hidden?.to_s != sub_category_params[:hidden] && sub_category.update(sub_category_params)
```
Comment (chu4) [inline app/controllers/account/admin/sub_categories_controller.rb:47]: `sub_category.hidden?.to_s != sub_category_params[:hidden]` can be moved to policy

File: app/controllers/account/admin/users_controller.rb [hunk 20]
```diff
@@ -69,5 +73,13 @@ def not_current_user
     def authorize_admin_user
       authorize current_user
     end
+
+    def set_page_title
+      @page_title = "Users"
+    end
+
+    def categories
```
Comment (chu4) [inline app/controllers/account/admin/users_controller.rb:81]: why do you need categories in users controller?

File: app/handlers/sub_cat_move_to_handler.rb [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+  def initialize(sub_category, params)
+    @sub_category = sub_category
+    @category_id = params[:move_to_cat_id]
+  end
+
+  def perform
+    if move_to_category.present? && sub_category.update(category: move_to_category)
+      current_category.sub_categories.where("position > ?", sub_category.position).update_all("position = position-1")
+      move_to_category.sub_categories.update_all("position = position+1")
+      move_to_category.sub_categories.where(id: sub_category.id).update(position: 1)
```
Comment (chu4) [inline app/handlers/sub_cat_move_to_handler.rb:13]: why not `sub_category.update(position: 1)`

File: spec/models/category_spec.rb [hunk 22]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Category, type: :model do
+  context "when valid category Factory" do
+    it "has a valid category factory" do
```
Comment (chu4) [inline spec/models/category_spec.rb:7]: useless spec already you already have single spec on validity of factories

File: spec/requests/account/admin/categories_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+        expect(response).to redirect_to(account_admin_categories_path)
+      end
+    end
+  end
+
+  describe '#update' do
+    context 'with valid params' do
+      before do
+        put account_admin_category_path(category),
+            params: { category: category_params.merge!(label: "Updated") }.merge(format: :js)
```
Comment (chu4) [inline spec/requests/account/admin/categories_controller_spec.rb:64]: why `merge format: :js` and not simply?
`params: { category: category_params.merge!(label: "Updated"), format: :js }`

File: spec/requests/account/admin/categories_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+      end
+
+      it { expect(assigns(:category)).to eq(category) }
+
+      it 'updates category attributes' do
+        expect(category.reload.label).to eq(category_params[:label])
```
Comment (chu4) [inline spec/requests/account/admin/categories_controller_spec.rb:70]: check if eq to "Updated"

File: spec/requests/account/admin/categories_controller_spec.rb (hunk 23 above)
Comment (chu4) [reply spec/requests/account/admin/categories_controller_spec.rb:64]: is there a need to override category_params with merge! ??

**Review by chu4 — APPROVED** [review]

**Pull request merged**